nx/lib/nx.ex

```diff
@@ -10375,6 +10375,198 @@ defmodule Nx do
+          ]
+        ]
+      >
+
+      iex> t = Nx.tensor([
+      ...>  [0, 1, 2],
+      ...>  [3, 4, 5],
+      ...>  [-1, -2, -3],
+      ...>  [-4, -5, -6],
+      ...>  [-7, -8, -9]
+      ...> ])
+      iex> Nx.as_windowed(t, {3, 2}, strides: [2, 1])
+      #Nx.Tensor<
+        s64[4][3][2]
+        [
+          [
+            [0, 1],
+            [3, 4],
+            [-1, -2]
+          ],
+          [
+            [1, 2],
+            [4, 5],
+            [-2, -3]
+          ],
+          [
+            [-1, -2],
+            [-4, -5],
+            [-7, -8]
+          ],
+          [
+            [-2, -3],
+            [-5, -6],
+            [-8, -9]
+          ]
+        ]
+      >
+  """
+  def as_windowed(tensor, window_dimensions, opts \\ []) do
+    %{shape: shape} = tensor = to_tensor(tensor)
+    opts = Keyword.validate!(opts, padding: :valid, strides: 1)
+
+    padding = opts[:padding]
+
+    strides =
+      case opts[:strides] do
+        strides when is_list(strides) ->
+          strides
+
+        strides when is_integer(strides) and strides >= 1 ->
+          List.duplicate(strides, Nx.rank(tensor))
+
+        strides ->
+          raise ArgumentError,
+                "expected an integer >= 1 or a list of integers, got: #{inspect(strides)}"
+      end
+
+    dilations = List.duplicate(1, Nx.rank(tensor))
+
+    {window_min_source_shape, _} =
+      Nx.Shape.pool(shape, window_dimensions, strides, padding, dilations)
+
+    window_min_source = Nx.broadcast(1, window_min_source_shape)
+
+    t_iota = Nx.iota(tensor)
+
+    # Scatter the min index for each window.
+    # Since strides is always != 0, we don't
+    # have any overlapping scatters, which
+    # results in a tensor which marks the start
+    # of each window.
+    scattered_indices =
+      t_iota
+      |> Nx.window_scatter_min(window_min_source, 0, window_dimensions,
+        strides: strides,
+        padding: padding
+      )
+      |> then(&Nx.select(&1, t_iota, -1))
+
+    # Since we can have windows which aren't touching,
+    # The can be "-1" gaps in the scattered_indices tensor
+    # above. For example:
+    # [
+    #   [0, 1, -1],
+    #   [-1, -1, -1],
+    #   [6, 7, -1],
+    #   [-1, -1, -1],
+    #   [-1, -1, -1]
+    # ]
+    # Because of this, we need to sort the tensor in all dimensions
+    # where the strides are bigger than 1. After sorting where needed,
+    # We need to slice the tensor to get all window starts in which we
+    # are interested.
+
+    zeros = List.duplicate(0, Nx.rank(scattered_indices))
+
+    window_dimensions_list = Tuple.to_list(window_dimensions)
+
+    sliced_indices =
+      for {stride, axis} <- Enum.with_index(strides), reduce: scattered_indices do
+        idx ->
+          len = elem(window_min_source.shape, axis)
+          lengths = idx.shape |> Tuple.to_list() |> List.replace_at(axis, len)
+
+          if stride > 1 do
+            sorted = Nx.sort(idx, axis: axis, direction: :asc)
```

Comment:
I can't convert this to a defn basically because I couldn't find a way to get the `axis` to increment in a `while`, for instance, and still keep it as a number.

Current defn implementation here: https://github.com/polvalente/nx-signal/blob/08c059d10d5bd90fcb433b2a5286e2a2e9441e99/lib/nx_signal.ex#L166